// 8004-solana/dist/core/indexer-graphql-client.js
/**
 * Indexer Client for GraphQL v2 API
 * Implements the IndexerReadClient contract used by the SDK.
 */
import { IndexerError, IndexerErrorCode, IndexerRateLimitError, IndexerTimeoutError, IndexerUnauthorizedError, IndexerUnavailableError, } from './indexer-errors.js';
const VALIDATION_ARCHIVED_ERROR = 'Validation feature is archived (v0.5.0+) and is not exposed by indexers.';
function toIsoFromUnixSeconds(unix) {
    if (typeof unix === 'string') {
        const trimmed = unix.trim();
        if (trimmed.length > 0 && !/^-?\d+(\.\d+)?$/.test(trimmed)) {
            const parsed = Date.parse(trimmed);
            if (Number.isFinite(parsed)) {
                return new Date(parsed).toISOString();
            }
        }
    }
    const n = typeof unix === 'string' ? Number(unix) : (typeof unix === 'number' ? unix : NaN);
    if (!Number.isFinite(n) || n <= 0)
        return new Date(0).toISOString();
    return new Date(n * 1000).toISOString();
}
function toNumberSafe(v, fallback = 0) {
    const n = typeof v === 'string' ? Number(v) : (typeof v === 'number' ? v : NaN);
    return Number.isFinite(n) ? n : fallback;
}
function toIntSafe(v, fallback = 0) {
    const n = typeof v === 'string' ? Number.parseInt(v, 10) : (typeof v === 'number' ? v : NaN);
    if (!Number.isFinite(n))
        return fallback;
    return Math.trunc(n);
}
function normalizeHexDigest(v) {
    if (typeof v !== 'string')
        return null;
    let s = v.trim();
    if (s.startsWith('\\x') || s.startsWith('0x'))
        s = s.slice(2);
    if (!s)
        return null;
    return s.toLowerCase();
}
function clampInt(n, min, max) {
    if (!Number.isFinite(n))
        return min;
    return Math.min(max, Math.max(min, Math.trunc(n)));
}
function normalizeGraphqlAgentLookupId(agentId) {
    if (typeof agentId === 'bigint') {
        if (agentId < 0n) {
            throw new IndexerError('agentId must be a non-negative integer or non-empty string', IndexerErrorCode.INVALID_RESPONSE);
        }
        return agentId.toString();
    }
    if (typeof agentId === 'number') {
        if (!Number.isFinite(agentId) || !Number.isInteger(agentId) || agentId < 0) {
            throw new IndexerError('agentId must be a non-negative integer or non-empty string', IndexerErrorCode.INVALID_RESPONSE);
        }
        return Math.trunc(agentId).toString();
    }
    const normalized = String(agentId).trim();
    if (!normalized) {
        throw new IndexerError('agentId must be a non-empty string or non-negative integer', IndexerErrorCode.INVALID_RESPONSE);
    }
    if (normalized.startsWith('sol:')) {
        const stripped = normalized.slice(4).trim();
        if (stripped)
            return stripped;
    }
    return normalized;
}
function toSafeGraphqlAgentIdNumber(agentId) {
    if (!/^\d+$/.test(agentId))
        return null;
    try {
        const parsed = BigInt(agentId);
        if (parsed > BigInt(Number.MAX_SAFE_INTEGER))
            return null;
        return Number(parsed);
    }
    catch {
        return null;
    }
}
function toGraphqlUnixSeconds(value) {
    if (value === undefined || value === null)
        return undefined;
    if (typeof value === 'number') {
        if (!Number.isFinite(value))
            return undefined;
        return Math.trunc(value).toString();
    }
    if (typeof value === 'string') {
        const trimmed = value.trim();
        if (!trimmed)
            return undefined;
        if (/^-?\d+$/.test(trimmed))
            return trimmed;
        const millis = Date.parse(trimmed);
        if (!Number.isFinite(millis))
            return undefined;
        return Math.floor(millis / 1000).toString();
    }
    if (value instanceof Date) {
        const millis = value.getTime();
        if (!Number.isFinite(millis))
            return undefined;
        return Math.floor(millis / 1000).toString();
    }
    return undefined;
}
function resolveAgentOrder(order) {
    const resolved = order ?? 'created_at.desc';
    const orderDirection = resolved.endsWith('.asc') ? 'asc' : 'desc';
    const field = resolved.split('.')[0] ?? 'created_at';
    const normalized = field.toLowerCase();
    if (normalized === 'updated_at' || normalized === 'updatedat') {
        return { orderBy: 'updatedAt', orderDirection };
    }
    if (normalized === 'total_feedback' || normalized === 'totalfeedback') {
        return { orderBy: 'totalFeedback', orderDirection };
    }
    if (normalized === 'quality_score' || normalized === 'qualityscore') {
        return { orderBy: 'qualityScore', orderDirection };
    }
    if (normalized === 'trust_tier' || normalized === 'trusttier') {
        return { orderBy: 'trustTier', orderDirection };
    }
    if (normalized === 'agentid' || normalized === 'agent_id') {
        // Keep compatibility for callers that still pass agent_id ordering.
        // GraphQL ordering support is backend-specific, so normalize to createdAt.
        return { orderBy: 'createdAt', orderDirection };
    }
    return { orderBy: 'createdAt', orderDirection };
}
function agentId(asset) {
    const normalized = asset.trim();
    if (normalized.startsWith('sol:')) {
        return normalized.slice(4);
    }
    return normalized;
}
function mapGqlAgent(agent, fallbackAsset = '') {
    const mappedAsset = agent?.solana?.assetPubkey ?? agent?.id ?? fallbackAsset;
    const mappedAgentId = agent?.agentid ?? agent?.agentId ?? agent?.globalId ?? agent?.global_id ?? null;
    return {
        agent_id: mappedAgentId,
        asset: mappedAsset,
        owner: agent?.owner ?? '',
        creator: agent?.creator ?? null,
        agent_uri: agent?.agentURI ?? null,
        agent_wallet: agent?.agentWallet ?? null,
        collection: agent?.solana?.collection ?? '',
        collection_pointer: agent?.collectionPointer ?? null,
        col_locked: Boolean(agent?.colLocked),
        parent_asset: agent?.parentAsset ?? null,
        parent_creator: agent?.parentCreator ?? null,
        parent_locked: Boolean(agent?.parentLocked),
        nft_name: null,
        atom_enabled: Boolean(agent?.solana?.atomEnabled),
        trust_tier: toNumberSafe(agent?.solana?.trustTier, 0),
        quality_score: toNumberSafe(agent?.solana?.qualityScore, 0),
        confidence: toNumberSafe(agent?.solana?.confidence, 0),
        risk_score: toNumberSafe(agent?.solana?.riskScore, 0),
        diversity_ratio: toNumberSafe(agent?.solana?.diversityRatio, 0),
        feedback_count: toNumberSafe(agent?.totalFeedback, 0),
        raw_avg_score: 0,
        sort_key: '0',
        block_slot: 0,
        tx_signature: '',
        created_at: toIsoFromUnixSeconds(agent?.createdAt),
        updated_at: toIsoFromUnixSeconds(agent?.updatedAt),
    };
}
function mapGqlCollectionPointer(row) {
    const collection = typeof row?.collection === 'string' ? row.collection : row?.col;
    const col = typeof row?.col === 'string' ? row.col : collection;
    const metadataUpdatedAt = row?.metadataUpdatedAt ?? row?.metadata_updated_at;
    return {
        collection: collection ?? col ?? '',
        col: col ?? collection ?? '',
        creator: row?.creator ?? '',
        first_seen_asset: row?.firstSeenAsset ?? row?.first_seen_asset ?? '',
        first_seen_at: toIsoFromUnixSeconds(row?.firstSeenAt ?? row?.first_seen_at),
        first_seen_slot: String(row?.firstSeenSlot ?? row?.first_seen_slot ?? '0'),
        first_seen_tx_signature: row?.firstSeenTxSignature ?? row?.first_seen_tx_signature ?? null,
        last_seen_at: toIsoFromUnixSeconds(row?.lastSeenAt ?? row?.last_seen_at),
        last_seen_slot: String(row?.lastSeenSlot ?? row?.last_seen_slot ?? '0'),
        last_seen_tx_signature: row?.lastSeenTxSignature ?? row?.last_seen_tx_signature ?? null,
        asset_count: String(row?.assetCount ?? row?.asset_count ?? '0'),
        version: row?.version ?? null,
        name: row?.name ?? null,
        symbol: row?.symbol ?? null,
        description: row?.description ?? null,
        image: row?.image ?? null,
        banner_image: row?.bannerImage ?? row?.banner_image ?? null,
        social_website: row?.socialWebsite ?? row?.social_website ?? null,
        social_x: row?.socialX ?? row?.social_x ?? null,
        social_discord: row?.socialDiscord ?? row?.social_discord ?? null,
        metadata_status: row?.metadataStatus ?? row?.metadata_status ?? null,
        metadata_hash: row?.metadataHash ?? row?.metadata_hash ?? null,
        metadata_bytes: row?.metadataBytes ?? row?.metadata_bytes ?? null,
        metadata_updated_at: metadataUpdatedAt !== undefined && metadataUpdatedAt !== null
            ? toIsoFromUnixSeconds(metadataUpdatedAt)
            : null,
    };
}
function buildAgentWhere(options) {
    if (!options)
        return {};
    const where = {};
    if (options.owner)
        where.owner = options.owner;
    if (options.creator)
        where.creator = options.creator;
    if (options.collection)
        where.collection = options.collection;
    if (options.collectionPointer)
        where.collectionPointer = options.collectionPointer;
    if (options.wallet)
        where.agentWallet = options.wallet;
    if (options.parentAsset)
        where.parentAsset = options.parentAsset;
    if (options.parentCreator)
        where.parentCreator = options.parentCreator;
    if (options.colLocked !== undefined)
        where.colLocked = options.colLocked;
    if (options.parentLocked !== undefined)
        where.parentLocked = options.parentLocked;
    const updatedAt = toGraphqlUnixSeconds(options.updatedAt);
    const updatedAtGt = toGraphqlUnixSeconds(options.updatedAtGt);
    const updatedAtLt = toGraphqlUnixSeconds(options.updatedAtLt);
    if (updatedAt !== undefined) {
        try {
            const exact = BigInt(updatedAt);
            where.updatedAt_gt = (exact - 1n).toString();
            where.updatedAt_lt = (exact + 1n).toString();
        }
        catch {
            // Ignore invalid numeric coercion and let explicit gt/lt (if any) drive the filter.
        }
    }
    if (updatedAtGt !== undefined)
        where.updatedAt_gt = updatedAtGt;
    if (updatedAtLt !== undefined)
        where.updatedAt_lt = updatedAtLt;
    return where;
}
function feedbackId(asset, client, index) {
    return `${asset}:${client}:${index.toString()}`;
}
function decodeFeedbackId(id) {
    const parts = id.split(':');
    if (parts.length === 3) {
        const [asset, client, index] = parts;
        if (asset === 'sol')
            return null;
        if (!asset || !client || !index)
            return null;
        return { asset, client, index };
    }
    if (parts.length === 4 && parts[0] === 'sol') {
        const [, asset, client, index] = parts;
        if (!asset || !client || !index)
            return null;
        return { asset, client, index };
    }
    return null;
}
function resolveFeedbackAsset(row, fallbackAsset = '') {
    if (typeof row?.id === 'string') {
        const decoded = decodeFeedbackId(row.id);
        if (decoded?.asset)
            return decoded.asset;
    }
    const directAgent = row?.agent;
    if (typeof directAgent === 'string' && directAgent.length > 0) {
        return directAgent;
    }
    if (typeof directAgent?.id === 'string' && directAgent.id.length > 0) {
        return directAgent.id;
    }
    if (typeof row?.asset === 'string' && row.asset.length > 0) {
        return row.asset;
    }
    return fallbackAsset;
}
function mapGqlFeedback(row, fallbackAsset = '') {
    return {
        id: row.id,
        asset: resolveFeedbackAsset(row, fallbackAsset),
        client_address: row.clientAddress,
        feedback_index: toNumberSafe(row.feedbackIndex, 0),
        value: row?.solana?.valueRaw ?? '0',
        value_decimals: toNumberSafe(row?.solana?.valueDecimals, 0),
        score: row?.solana?.score ?? null,
        tag1: row.tag1 ?? '',
        tag2: row.tag2 ?? '',
        endpoint: row.endpoint ?? null,
        feedback_uri: row.feedbackURI ?? null,
        running_digest: null,
        feedback_hash: row.feedbackHash ?? null,
        is_revoked: Boolean(row.isRevoked),
        revoked_at: row.revokedAt ? toIsoFromUnixSeconds(row.revokedAt) : null,
        block_slot: toNumberSafe(row?.solana?.blockSlot, 0),
        tx_signature: row?.solana?.txSignature ?? '',
        created_at: toIsoFromUnixSeconds(row.createdAt),
    };
}
function mapGqlFeedbackResponse(row, asset, client, feedbackIndex) {
    return {
        id: row.id,
        asset,
        client_address: client,
        feedback_index: toNumberSafe(feedbackIndex, 0),
        responder: row.responder,
        response_uri: row.responseUri ?? null,
        response_hash: row.responseHash ?? null,
        running_digest: null,
        block_slot: toNumberSafe(row?.solana?.blockSlot, 0),
        tx_signature: row?.solana?.txSignature ?? '',
        created_at: toIsoFromUnixSeconds(row.createdAt),
    };
}
export class IndexerGraphQLClient {
    graphqlUrl;
    headers;
    timeout;
    retries;
    hashChainHeadsInFlight = new Map();
    constructor(config) {
        this.graphqlUrl = config.graphqlUrl.replace(/\/$/, '');
        this.headers = config.headers ?? {};
        this.timeout = config.timeout ?? 10000;
        this.retries = config.retries ?? 2;
    }
    getBaseUrl() {
        return this.graphqlUrl;
    }
    shouldUseLegacyCollectionRead(error) {
        if (!(error instanceof IndexerError))
            return false;
        if (error.code !== IndexerErrorCode.INVALID_RESPONSE)
            return false;
        const msg = error.message;
        return (/Cannot query field ['"]collections['"] on type ['"]Query['"]/.test(msg)
            || /Unknown argument ['"]collection['"] on field ['"]Query\.collectionAssetCount['"]/.test(msg)
            || /Unknown argument ['"]collection['"] on field ['"]Query\.collectionAssets['"]/.test(msg));
    }
    shouldFallbackAgentIdField(error, field) {
        if (!(error instanceof IndexerError))
            return false;
        if (error.code !== IndexerErrorCode.INVALID_RESPONSE)
            return false;
        const msg = error.message;
        return (new RegExp(`Cannot query field ['"]${field}['"] on type ['"]Agent['"]`).test(msg)
            || new RegExp(`Cannot query field ['"]${field}['"] on type ['"]AgentFilter['"]`).test(msg)
            || new RegExp(`Field ['"]${field}['"] is not defined by type ['"]AgentFilter['"]`).test(msg)
            || new RegExp(`Unknown argument ['"]${field}['"]`).test(msg)
            || new RegExp(`Unknown field ['"]${field}['"]`).test(msg));
    }
    shouldFallbackAgentIdVariableType(error, variableType) {
        if (!(error instanceof IndexerError))
            return false;
        if (error.code !== IndexerErrorCode.INVALID_RESPONSE)
            return false;
        const msg = error.message;
        if (variableType === 'String') {
            return (/type ['"]String!?['"] used in position expecting type ['"]BigInt!?['"]/i.test(msg)
                || /Expected type ['"]BigInt!?['"]/i.test(msg)
                || /expecting type ['"]BigInt!?['"]/i.test(msg));
        }
        return false;
    }
    shouldRetryBigIntAgentIdAsNumber(error) {
        if (!(error instanceof IndexerError))
            return false;
        if (error.code !== IndexerErrorCode.INVALID_RESPONSE)
            return false;
        const msg = error.message;
        return (/BigInt cannot represent non-integer value/i.test(msg)
            || /Expected value of type ['"]BigInt!?['"], found ['"][^'"]+['"]/i.test(msg)
            || /Expected type ['"]BigInt!?['"], found ['"][^'"]+['"]/i.test(msg));
    }
    async requestAgentBySequentialIdField(agentIdField, normalizedAgentId) {
        const requestByType = async (variableType, variableValue) => {
            const data = await this.request(`query($agentId: ${variableType}!) {
          agents(first: 1, where: { ${agentIdField}: $agentId }) {
            id
            owner
            creator
            agentURI
            agentWallet
            collectionPointer
            colLocked
            parentAsset
            parentCreator
            parentLocked
            createdAt
            updatedAt
            totalFeedback
            solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
          }
        }`, { agentId: variableValue });
            return data.agents[0] ?? null;
        };
        try {
            return await requestByType('String', normalizedAgentId);
        }
        catch (error) {
            if (!this.shouldFallbackAgentIdVariableType(error, 'String')) {
                throw error;
            }
        }
        try {
            return await requestByType('BigInt', normalizedAgentId);
        }
        catch (error) {
            const safeNumericAgentId = toSafeGraphqlAgentIdNumber(normalizedAgentId);
            if (safeNumericAgentId !== null && this.shouldRetryBigIntAgentIdAsNumber(error)) {
                return requestByType('BigInt', safeNumericAgentId);
            }
            throw error;
        }
    }
    async requestWithAgentIdField(requester) {
        try {
            return await requester('agentId');
        }
        catch (error) {
            if (!this.shouldFallbackAgentIdField(error, 'agentId')) {
                throw error;
            }
        }
        try {
            return await requester('agentid');
        }
        catch (error) {
            if (!this.shouldFallbackAgentIdField(error, 'agentid')) {
                throw error;
            }
        }
        return requester(null);
    }
    async request(query, variables) {
        let lastError = null;
        for (let attempt = 0; attempt <= this.retries; attempt++) {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);
            try {
                const response = await fetch(this.graphqlUrl, {
                    method: 'POST',
                    headers: {
                        'content-type': 'application/json',
                        ...this.headers,
                    },
                    body: JSON.stringify({ query, variables }),
                    signal: controller.signal,
                    redirect: 'error',
                });
                if (!response.ok) {
                    // Many GraphQL servers (including ours) can return JSON error bodies with HTTP 400.
                    // Surface those messages to help diagnose query complexity/validation issues.
                    let details = '';
                    try {
                        const contentType = response.headers.get('content-type') ?? '';
                        const text = await response.text();
                        if (contentType.includes('application/json')) {
                            const parsed = JSON.parse(text);
                            const msg = parsed?.errors?.map(e => e?.message).filter(Boolean).join('; ');
                            if (msg)
                                details = msg;
                        }
                        else if (text) {
                            details = text.slice(0, 200).replace(/\s+/g, ' ').trim();
                        }
                    }
                    catch {
                        // Ignore body parsing issues for non-OK responses.
                    }
                    if (response.status === 401 || response.status === 403) {
                        throw new IndexerUnauthorizedError();
                    }
                    if (response.status === 429) {
                        const retryAfter = response.headers.get('Retry-After');
                        throw new IndexerRateLimitError('Rate limited', retryAfter ? parseInt(retryAfter, 10) : undefined);
                    }
                    if (attempt < this.retries) {
                        await new Promise(r => setTimeout(r, 100 * Math.pow(2, attempt)));
                        continue;
                    }
                    throw new IndexerError(`GraphQL request failed: HTTP ${response.status}${details ? ` (${details})` : ''}`, IndexerErrorCode.SERVER_ERROR);
                }
                const json = (await response.json());
                if (json.errors && json.errors.length > 0) {
                    const msg = json.errors.map(e => e?.message).filter(Boolean).join('; ') || 'GraphQL error';
                    throw new IndexerError(msg, IndexerErrorCode.INVALID_RESPONSE);
                }
                if (!json.data) {
                    throw new IndexerError('GraphQL response missing data', IndexerErrorCode.INVALID_RESPONSE);
                }
                return json.data;
            }
            catch (err) {
                const e = err;
                lastError = err instanceof Error ? err : new Error(String(err));
                if (e?.name === 'AbortError') {
                    lastError = new IndexerTimeoutError();
                }
                else if (!(err instanceof IndexerError)) {
                    // Network / fetch errors
                    if (err instanceof TypeError) {
                        lastError = new IndexerUnavailableError(err.message);
                    }
                }
                if (attempt < this.retries) {
                    await new Promise(r => setTimeout(r, 100 * Math.pow(2, attempt)));
                    continue;
                }
                throw lastError instanceof IndexerError ? lastError : new IndexerUnavailableError(lastError.message);
            }
            finally {
                clearTimeout(timeoutId);
            }
        }
        throw lastError ?? new IndexerUnavailableError();
    }
    async isAvailable() {
        try {
            await this.request('query { __typename }');
            return true;
        }
        catch {
            return false;
        }
    }
    loadHashChainHeads(asset) {
        const key = asset;
        const existing = this.hashChainHeadsInFlight.get(key);
        if (existing)
            return existing;
        const pending = this.request(`query($agent: ID!) {
        hashChainHeads(agent: $agent) {
          feedback { digest count }
          response { digest count }
          revoke { digest count }
        }
      }`, { agent: agentId(asset) })
            .then((d) => d.hashChainHeads)
            .finally(() => {
            this.hashChainHeadsInFlight.delete(key);
        });
        this.hashChainHeadsInFlight.set(key, pending);
        return pending;
    }
    // ============================================================================
    // Agents
    // ============================================================================
    async getAgent(asset) {
        const normalizedAsset = agentId(asset);
        const data = await this.requestWithAgentIdField((agentIdField) => {
            const agentIdSelection = agentIdField ? `\n          ${agentIdField}` : '';
            return this.request(`query($id: ID!) {
            agent(id: $id) {
              id${agentIdSelection}
              owner
              creator
              agentURI
              agentWallet
              collectionPointer
              colLocked
              parentAsset
              parentCreator
              parentLocked
              createdAt
              updatedAt
              totalFeedback
              solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
            }
          }`, { id: normalizedAsset });
        });
        if (!data.agent)
            return null;
        return mapGqlAgent(data.agent, normalizedAsset);
    }
    async getAgentByAgentId(agentId) {
        const normalizedAgentId = normalizeGraphqlAgentLookupId(agentId);
        const agent = await this.requestWithAgentIdField(async (agentIdField) => {
            if (agentIdField === null) {
                const legacy = await this.request(`query($id: ID!) {
              agent(id: $id) {
                id
                owner
                creator
                agentURI
                agentWallet
                collectionPointer
                colLocked
                parentAsset
                parentCreator
                parentLocked
                createdAt
                updatedAt
                totalFeedback
                solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
              }
            }`, { id: normalizedAgentId });
                return legacy.agent;
            }
            return this.requestAgentBySequentialIdField(agentIdField, normalizedAgentId);
        });
        if (!agent)
            return null;
        const mapped = mapGqlAgent(agent, normalizedAgentId);
        mapped.agent_id = normalizedAgentId;
        return mapped;
    }
    /** @deprecated Use getAgentByAgentId(agentId) */
    async getAgentByIndexerId(agentId) {
        return this.getAgentByAgentId(agentId);
    }
    async getAgents(options) {
        const limit = clampInt(options?.limit ?? 100, 0, 500);
        const offset = clampInt(options?.offset ?? 0, 0, 1_000_000);
        const { orderBy, orderDirection } = resolveAgentOrder(options?.order);
        const where = buildAgentWhere(options);
        const data = await this.requestWithAgentIdField((agentIdField) => {
            const agentIdSelection = agentIdField ? `\n          ${agentIdField}` : '';
            return this.request(`query($orderBy: AgentOrderBy!, $dir: OrderDirection!, $where: AgentFilter) {
            agents(first: ${limit}, skip: ${offset}, where: $where, orderBy: $orderBy, orderDirection: $dir) {
              id${agentIdSelection}
              owner
              creator
              agentURI
              agentWallet
              collectionPointer
              colLocked
              parentAsset
              parentCreator
              parentLocked
              createdAt
              updatedAt
              totalFeedback
              solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
            }
          }`, {
                orderBy,
                dir: orderDirection,
                where: Object.keys(where).length ? where : null,
            });
        });
        return data.agents.map((a) => mapGqlAgent(a));
    }
    async getAgentsByOwner(owner) {
        return this.getAgents({
            owner,
            limit: 250,
            order: 'created_at.desc',
        });
    }
    async getAgentsByCollection(collection) {
        return this.getAgents({
            collection,
            limit: 250,
            order: 'created_at.desc',
        });
    }
    async getAgentByWallet(wallet) {
        const agents = await this.getAgents({
            wallet,
            limit: 1,
            order: 'created_at.desc',
        });
        return agents[0] ?? null;
    }
    async getLeaderboard(options) {
        if (options?.cursorSortKey) {
            throw new Error('GraphQL backend does not support cursorSortKey keyset pagination; use REST indexer client.');
        }
        const limit = clampInt(options?.limit ?? 50, 0, 200);
        const where = {};
        if (options?.collection)
            where.collection = options.collection;
        if (options?.minTier !== undefined)
            where.trustTier_gte = options.minTier;
        const data = await this.requestWithAgentIdField((agentIdField) => {
            const agentIdSelection = agentIdField ? `${agentIdField} ` : '';
            return this.request(`query($where: AgentFilter) {
            agents(first: ${limit}, where: $where, orderBy: qualityScore, orderDirection: desc) {
              ${agentIdSelection}owner creator agentURI agentWallet collectionPointer colLocked parentAsset parentCreator parentLocked createdAt updatedAt totalFeedback
              solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
            }
          }`, { where: Object.keys(where).length ? where : null });
        });
        return data.agents.map((a) => mapGqlAgent(a));
    }
    async getGlobalStats() {
        const data = await this.request(`query {
        globalStats { totalAgents totalFeedback totalCollections tags }
      }`);
        const stats = data.globalStats;
        return {
            total_agents: toNumberSafe(stats?.totalAgents, 0),
            total_collections: toNumberSafe(stats?.totalCollections, 0),
            total_feedbacks: toNumberSafe(stats?.totalFeedback, 0),
            total_validations: 0,
            platinum_agents: 0,
            gold_agents: 0,
            avg_quality: null,
        };
    }
    async getCollectionPointers(options) {
        const first = clampInt(options?.limit ?? 100, 0, 500);
        const skip = clampInt(options?.offset ?? 0, 0, 1_000_000);
        const collection = options?.collection ?? options?.col;
        try {
            const data = await this.request(`query($first: Int!, $skip: Int!, $collection: String, $creator: String) {
          collections(first: $first, skip: $skip, collection: $collection, creator: $creator) {
            collection
            creator
            firstSeenAsset
            firstSeenAt
            firstSeenSlot
            firstSeenTxSignature
            lastSeenAt
            lastSeenSlot
            lastSeenTxSignature
            assetCount
            version
            name
            symbol
            description
            image
            bannerImage
            socialWebsite
            socialX
            socialDiscord
            metadataStatus
            metadataHash
            metadataBytes
            metadataUpdatedAt
          }
        }`, {
                first,
                skip,
                collection: collection ?? null,
                creator: options?.creator ?? null,
            });
            return data.collections.map((p) => mapGqlCollectionPointer(p));
        }
        catch (error) {
            if (!this.shouldUseLegacyCollectionRead(error)) {
                throw error;
            }
            const data = await this.request(`query($first: Int!, $skip: Int!, $col: String, $creator: String) {
          collectionPointers(first: $first, skip: $skip, col: $col, creator: $creator) {
            col
            creator
            firstSeenAsset
            firstSeenAt
            firstSeenSlot
            firstSeenTxSignature
            lastSeenAt
            lastSeenSlot
            lastSeenTxSignature
            assetCount
          }
        }`, {
                first,
                skip,
                col: collection ?? null,
                creator: options?.creator ?? null,
            });
            return data.collectionPointers.map((p) => mapGqlCollectionPointer(p));
        }
    }
    async getCollectionAssetCount(col, creator) {
        try {
            const data = await this.request(`query($collection: String!, $creator: String) {
          collectionAssetCount(collection: $collection, creator: $creator)
        }`, {
                collection: col,
                creator: creator ?? null,
            });
            return toIntSafe(data.collectionAssetCount, 0);
        }
        catch (error) {
            if (!this.shouldUseLegacyCollectionRead(error)) {
                throw error;
            }
            const data = await this.request(`query($col: String!, $creator: String) {
          collectionAssetCount(col: $col, creator: $creator)
        }`, {
                col,
                creator: creator ?? null,
            });
            return toIntSafe(data.collectionAssetCount, 0);
        }
    }
    async getCollectionAssets(col, options) {
        const first = clampInt(options?.limit ?? 100, 0, 500);
        const skip = clampInt(options?.offset ?? 0, 0, 1_000_000);
        const order = options?.order ?? 'created_at.desc';
        const orderDirection = order.includes('.asc') ? 'asc' : 'desc';
        const orderBy = order.startsWith('updated_at')
            ? 'updatedAt'
            : order.startsWith('total_feedback')
                ? 'totalFeedback'
                : order.startsWith('quality_score')
                    ? 'qualityScore'
                    : order.startsWith('trust_tier')
                        ? 'trustTier'
                        : 'createdAt';
        return this.requestWithAgentIdField(async (agentIdField) => {
            const agentIdSelection = agentIdField ? `\n            ${agentIdField}` : '';
            try {
                const data = await this.request(`query($collection: String!, $creator: String, $first: Int!, $skip: Int!, $orderBy: AgentOrderBy!, $dir: OrderDirection!) {
            collectionAssets(
              collection: $collection,
              creator: $creator,
              first: $first,
              skip: $skip,
              orderBy: $orderBy,
              orderDirection: $dir
            ) {
              id${agentIdSelection}
              owner
              creator
              agentURI
              agentWallet
              collectionPointer
              colLocked
              parentAsset
              parentCreator
              parentLocked
              createdAt
              updatedAt
              totalFeedback
              solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
            }
          }`, {
                    collection: col,
                    creator: options?.creator ?? null,
                    first,
                    skip,
                    orderBy,
                    dir: orderDirection,
                });
                return data.collectionAssets.map((a) => mapGqlAgent(a));
            }
            catch (error) {
                if (!this.shouldUseLegacyCollectionRead(error)) {
                    throw error;
                }
                const data = await this.request(`query($col: String!, $creator: String, $first: Int!, $skip: Int!, $orderBy: AgentOrderBy!, $dir: OrderDirection!) {
            collectionAssets(
              col: $col,
              creator: $creator,
              first: $first,
              skip: $skip,
              orderBy: $orderBy,
              orderDirection: $dir
            ) {
              id${agentIdSelection}
              owner
              creator
              agentURI
              agentWallet
              collectionPointer
              colLocked
              parentAsset
              parentCreator
              parentLocked
              createdAt
              updatedAt
              totalFeedback
              solana { assetPubkey collection atomEnabled trustTier qualityScore confidence riskScore diversityRatio }
            }
          }`, {
                    col,
                    creator: options?.creator ?? null,
                    first,
                    skip,
                    orderBy,
                    dir: orderDirection,
                });
                return data.collectionAssets.map((a) => mapGqlAgent(a));
            }
        });
    }
    // ============================================================================
    // Feedbacks
    // ============================================================================
    async getFeedbacks(asset, options) {
        const limit = clampInt(options?.limit ?? 100, 0, 1000);
        const initialSkip = clampInt(options?.offset ?? 0, 0, 1_000_000);
        if (limit === 0)
            return [];
        const where = { agent: agentId(asset) };
        if (!options?.includeRevoked) {
            where.isRevoked = false;
        }
        const pageSize = 100;
        const feedbacks = [];
        let skip = initialSkip;
        while (feedbacks.length < limit) {
            const first = Math.min(pageSize, limit - feedbacks.length);
            const data = await this.request(`query($where: FeedbackFilter) {
          feedbacks(first: ${first}, skip: ${skip}, where: $where, orderBy: createdAt, orderDirection: desc) {
            id
            clientAddress
            feedbackIndex
            tag1
            tag2
            endpoint
            feedbackURI
            feedbackHash
            isRevoked
            createdAt
            revokedAt
            solana { valueRaw valueDecimals score txSignature blockSlot }
          }
        }`, { where });
            const page = data.feedbacks.map((f) => mapGqlFeedback(f, asset));
            if (page.length === 0)
                break;
            feedbacks.push(...page);
            skip += page.length;
        }
        return feedbacks;
    }
    async getFeedback(asset, client, feedbackIndex) {
        const data = await this.request(`query($id: ID!) {
        feedback(id: $id) {
          id
          clientAddress
          feedbackIndex
          tag1
          tag2
          endpoint
          feedbackURI
      feedbackHash
      isRevoked
      createdAt
      revokedAt
      solana { valueRaw valueDecimals score txSignature blockSlot }
    }
  }`, { id: feedbackId(asset, client, feedbackIndex) });
        if (!data.feedback)
            return null;
        return mapGqlFeedback(data.feedback, asset);
    }
    async getFeedbacksByClient(client) {
        const data = await this.request(`query($client: String!) {
        feedbacks(first: 250, where: { clientAddress: $client }, orderBy: createdAt, orderDirection: desc) {
          id
          agent { id }
          clientAddress
          feedbackIndex
          tag1
          tag2
          endpoint
          feedbackURI
          feedbackHash
          isRevoked
          createdAt
          revokedAt
          solana { valueRaw valueDecimals score txSignature blockSlot }
        }
      }`, { client });
        return data.feedbacks.map((f) => mapGqlFeedback(f));
    }
    async getFeedbacksByTag(tag) {
        // GraphQL filter doesn't support OR on tag1/tag2, so query both and merge.
        // Use paginated reads to stay below hosted GraphQL complexity limits.
        const pageSize = 100;
        const maxRows = 5000;
        const fetchByTagField = async (field) => {
            const rows = [];
            let skip = 0;
            while (rows.length < maxRows) {
                const first = Math.min(pageSize, maxRows - rows.length);
                const data = await this.request(`query($tag: String!) {
            feedbacks(
              first: ${first},
              skip: ${skip},
              where: { ${field}: $tag },
              orderBy: createdAt,
              orderDirection: desc
            ) {
              id
              agent { id }
              clientAddress
              feedbackIndex
              tag1
              tag2
              endpoint
              feedbackURI
              feedbackHash
              isRevoked
              createdAt
              revokedAt
              solana { valueRaw valueDecimals score txSignature blockSlot }
            }
          }`, { tag });
                const page = data.feedbacks ?? [];
                if (page.length === 0)
                    break;
                rows.push(...page);
                if (page.length < first)
                    break;
                skip += page.length;
            }
            return rows;
        };
        const [tag1Rows, tag2Rows] = await Promise.all([
            fetchByTagField('tag1'),
            fetchByTagField('tag2'),
        ]);
        const merged = new Map();
        for (const f of [...tag1Rows, ...tag2Rows]) {
            merged.set(f.id, f);
        }
        return Array.from(merged.values()).map((f) => mapGqlFeedback(f));
    }
    async getFeedbacksByEndpoint(endpoint) {
        const pageSize = 100;
        const maxRows = 5000;
        const rows = [];
        let skip = 0;
        while (rows.length < maxRows) {
            const first = Math.min(pageSize, maxRows - rows.length);
            const data = await this.request(`query($endpoint: String!) {
          feedbacks(
            first: ${first},
            skip: ${skip},
            where: { endpoint: $endpoint },
            orderBy: createdAt,
            orderDirection: desc
          ) {
            id
            agent { id }
            clientAddress
            feedbackIndex
            tag1
            tag2
            endpoint
            feedbackURI
            feedbackHash
            isRevoked
            createdAt
            revokedAt
            solana { valueRaw valueDecimals score txSignature blockSlot }
          }
        }`, { endpoint });
            const page = data.feedbacks ?? [];
            if (page.length === 0)
                break;
            rows.push(...page);
            if (page.length < first)
                break;
            skip += page.length;
        }
        return rows.map((f) => mapGqlFeedback(f));
    }
    async getAllFeedbacks(options) {
        const first = clampInt(options?.limit ?? 5000, 0, 5000);
        const where = {};
        if (!options?.includeRevoked)
            where.isRevoked = false;
        const data = await this.request(`query($where: FeedbackFilter) {
        feedbacks(first: ${first}, where: $where, orderBy: createdAt, orderDirection: desc) {
          id
          agent { id }
          clientAddress
          feedbackIndex
          tag1
          tag2
          endpoint
          feedbackURI
          feedbackHash
          isRevoked
          createdAt
          revokedAt
          solana { valueRaw valueDecimals score txSignature blockSlot }
        }
      }`, { where: Object.keys(where).length ? where : null });
        return data.feedbacks.map((f) => mapGqlFeedback(f));
    }
    async getLastFeedbackIndex(asset, client) {
        const data = await this.request(`query($agent: ID!, $client: String!) {
        feedbacks(first: 1, where: { agent: $agent, clientAddress: $client }, orderBy: feedbackIndex, orderDirection: desc) {
          feedbackIndex
        }
      }`, { agent: agentId(asset), client });
        if (!data.feedbacks || data.feedbacks.length === 0)
            return -1n;
        return BigInt(data.feedbacks[0].feedbackIndex);
    }
    // ============================================================================
    // Responses
    // ============================================================================
    async getFeedbackResponsesFor(asset, client, feedbackIndex, limit = 100) {
        const data = await this.request(`query($feedback: ID!) {
        feedbackResponses(first: ${clampInt(limit, 0, 1000)}, where: { feedback: $feedback }, orderBy: createdAt, orderDirection: asc) {
          id
          responder
          responseUri
          responseHash
          createdAt
          solana { txSignature blockSlot }
        }
      }`, { feedback: feedbackId(asset, client, feedbackIndex) });
        return (data.feedbackResponses ?? []).map((r) => mapGqlFeedbackResponse(r, asset, client, feedbackIndex));
    }
    // ============================================================================
    // Validations
    // ============================================================================
    async getPendingValidations(_validator) {
        throw new Error(VALIDATION_ARCHIVED_ERROR);
    }
    // ============================================================================
    // Reputation
    // ============================================================================
    async getAgentReputation(asset) {
        const normalizedAsset = agentId(asset);
        const data = await this.request(`query($id: ID!) {
        agent(id: $id) {
          id
          owner
          agentURI
          totalFeedback
          solana { assetPubkey collection qualityScore }
        }
      }`, { id: normalizedAsset });
        const row = data.agent;
        if (!row)
            return null;
        const feedbackCount = Math.max(0, toIntSafe(row?.totalFeedback, 0));
        const rawQualityScore = toNumberSafe(row?.solana?.qualityScore, 0);
        const qualityAvg = Math.max(0, Math.min(100, rawQualityScore > 100 ? rawQualityScore / 100 : rawQualityScore));
        const scores = [];
        const maxRows = Math.min(feedbackCount, 5000);
        const pageSize = 100;
        let skip = 0;
        while (skip < maxRows) {
            const first = Math.min(pageSize, maxRows - skip);
            const pageData = await this.request(`query($agent: ID!) {
          feedbacks(
            first: ${first},
            skip: ${skip},
            where: { agent: $agent },
            orderBy: createdAt,
            orderDirection: desc
          ) {
            solana { score }
          }
        }`, { agent: normalizedAsset });
            const page = pageData.feedbacks ?? [];
            if (page.length === 0)
                break;
            for (const feedback of page) {
                const score = feedback?.solana?.score;
                if (score === null || score === undefined)
                    continue;
                const parsed = toNumberSafe(score, Number.NaN);
                if (Number.isFinite(parsed)) {
                    scores.push(parsed);
                }
            }
            skip += page.length;
            if (page.length < first)
                break;
        }
        let avgScore = feedbackCount > 0 ? qualityAvg : null;
        let positiveCount = 0;
        let negativeCount = 0;
        if (scores.length > 0) {
            const sum = scores.reduce((acc, score) => acc + score, 0);
            avgScore = sum / scores.length;
            positiveCount = scores.filter((score) => score >= 50).length;
            negativeCount = scores.length - positiveCount;
        }
        const observedCount = positiveCount + negativeCount;
        if (feedbackCount > observedCount) {
            const remaining = feedbackCount - observedCount;
            const ratio = avgScore === null ? 0 : Math.max(0, Math.min(1, avgScore / 100));
            const estimatedPositive = Math.round(remaining * ratio);
            positiveCount += estimatedPositive;
            negativeCount += remaining - estimatedPositive;
        }
        return {
            asset: row?.solana?.assetPubkey ?? row?.id ?? normalizedAsset,
            owner: row?.owner ?? '',
            collection: row?.solana?.collection ?? '',
            nft_name: null,
            agent_uri: row?.agentURI ?? null,
            feedback_count: feedbackCount,
            avg_score: feedbackCount > 0 ? avgScore : null,
            positive_count: positiveCount,
            negative_count: negativeCount,
            validation_count: 0,
        };
    }
    // ============================================================================
    // Integrity (hash-chain)
    // ============================================================================
    async getLastFeedbackDigest(asset) {
        const heads = await this.loadHashChainHeads(asset);
        return {
            digest: normalizeHexDigest(heads.feedback.digest),
            count: toIntSafe(heads.feedback.count, 0),
        };
    }
    async getLastResponseDigest(asset) {
        const heads = await this.loadHashChainHeads(asset);
        return {
            digest: normalizeHexDigest(heads.response.digest),
            count: toIntSafe(heads.response.count, 0),
        };
    }
    async getLastRevokeDigest(asset) {
        const heads = await this.loadHashChainHeads(asset);
        return {
            digest: normalizeHexDigest(heads.revoke.digest),
            count: toIntSafe(heads.revoke.count, 0),
        };
    }
    async getLatestCheckpoints(asset) {
        const data = await this.request(`query($agent: ID!) {
        hashChainLatestCheckpoints(agent: $agent) {
          feedback { eventCount digest createdAt }
          response { eventCount digest createdAt }
          revoke { eventCount digest createdAt }
        }
      }`, { agent: agentId(asset) });
        const mapCp = (cp) => {
            if (!cp)
                return null;
            return {
                event_count: toIntSafe(cp.eventCount, 0),
                digest: normalizeHexDigest(cp.digest) ?? cp.digest,
                created_at: toIsoFromUnixSeconds(cp.createdAt),
            };
        };
        return {
            feedback: mapCp(data.hashChainLatestCheckpoints.feedback),
            response: mapCp(data.hashChainLatestCheckpoints.response),
            revoke: mapCp(data.hashChainLatestCheckpoints.revoke),
        };
    }
    async getReplayData(asset, chainType, fromCount = 0, toCount = 1000, limit = 1000) {
        const first = clampInt(limit, 1, 1000);
        const data = await this.request(`query($agent: ID!, $chainType: HashChainType!, $fromCount: BigInt!, $toCount: BigInt, $first: Int!) {
        hashChainReplayData(
          agent: $agent,
          chainType: $chainType,
          fromCount: $fromCount,
          toCount: $toCount,
          first: $first
        ) {
          hasMore
          nextFromCount
          events {
            asset
            client
            feedbackIndex
            slot
            runningDigest
            feedbackHash
            responder
            responseHash
            responseCount
            revokeCount
          }
        }
      }`, {
            agent: agentId(asset),
            chainType: chainType.toUpperCase(),
            fromCount: String(fromCount),
            toCount: toCount != null ? String(toCount) : null,
            first,
        });
        const page = data.hashChainReplayData;
        const events = page.events.map((e) => ({
            asset: e.asset,
            client: e.client,
            feedback_index: String(e.feedbackIndex),
            slot: toNumberSafe(e.slot, 0),
            running_digest: normalizeHexDigest(e.runningDigest) ?? null,
            feedback_hash: e.feedbackHash ?? null,
            responder: e.responder ?? undefined,
            response_hash: e.responseHash ?? null,
            response_count: e.responseCount != null ? toNumberSafe(e.responseCount, 0) : null,
            revoke_count: e.revokeCount != null ? toNumberSafe(e.revokeCount, 0) : null,
        }));
        return {
            events,
            hasMore: Boolean(page.hasMore),
            nextFromCount: toNumberSafe(page.nextFromCount, fromCount),
        };
    }
    async getFeedbacksAtIndices(asset, indices) {
        const result = new Map();
        if (indices.length === 0)
            return result;
        for (const idx of indices) {
            result.set(idx, null);
        }
        await Promise.all(indices.map(async (idx) => {
            const page = await this.getReplayData(asset, 'feedback', idx, idx + 1, 1);
            const e = page.events[0];
            if (!e)
                return;
            result.set(idx, {
                id: '',
                asset,
                client_address: e.client,
                feedback_index: toIntSafe(e.feedback_index, 0),
                value: '0',
                value_decimals: 0,
                score: null,
                tag1: null,
                tag2: null,
                endpoint: null,
                feedback_uri: null,
                running_digest: e.running_digest,
                feedback_hash: e.feedback_hash ?? null,
                is_revoked: false,
                revoked_at: null,
                block_slot: e.slot,
                tx_signature: '',
                created_at: new Date(0).toISOString(),
            });
        }));
        return result;
    }
    async getResponsesAtOffsets(asset, offsets) {
        const result = new Map();
        if (offsets.length === 0)
            return result;
        for (const offset of offsets) {
            result.set(offset, null);
        }
        await Promise.all(offsets.map(async (offset) => {
            const page = await this.getReplayData(asset, 'response', offset, offset + 1, 1);
            const e = page.events[0];
            if (!e)
                return;
            result.set(offset, {
                id: '',
                asset,
                client_address: e.client,
                feedback_index: toIntSafe(e.feedback_index, 0),
                responder: e.responder ?? '',
                response_uri: null,
                response_hash: e.response_hash ?? null,
                running_digest: e.running_digest,
                block_slot: e.slot,
                tx_signature: '',
                created_at: new Date(0).toISOString(),
            });
        }));
        return result;
    }
    async getRevocationsAtCounts(asset, revokeCounts) {
        const result = new Map();
        if (revokeCounts.length === 0)
            return result;
        for (const c of revokeCounts) {
            result.set(c, null);
        }
        await Promise.all(revokeCounts.map(async (c) => {
            if (!Number.isFinite(c) || c < 1)
                return;
            const idx = c - 1;
            const page = await this.getReplayData(asset, 'revoke', idx, idx + 1, 1);
            const e = page.events[0];
            if (!e)
                return;
            result.set(c, {
                id: '',
                asset,
                client_address: e.client,
                feedback_index: toIntSafe(e.feedback_index, 0),
                feedback_hash: e.feedback_hash ?? null,
                slot: e.slot,
                original_score: null,
                atom_enabled: false,
                had_impact: false,
                running_digest: e.running_digest,
                revoke_count: e.revoke_count ?? idx,
                tx_signature: '',
                created_at: new Date(0).toISOString(),
            });
        }));
        return result;
    }
}
